feat: add http datasource

teddylear opened the pull request:
## Description
Creating http datasource similar to [terraform http data source](https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http).

## Testing
Commands to run tests: 
```bash
make 
PACKER_ACC=1 go test -count 1 -v ./datasource/http/data_acc_test.go  -timeout=2m
```

Closes #11552

teddylear commented:
@nywilken Looks like this needs Vercel permission, is that something you can please grant so CI can run. If there is something I need to run on my end please let me know. Thanks!

teddylear commented:
@nywilken @azr Is there something I'm missing to trigger the markdown file builds? Or should I just push another commit to get this firing?

sylviamoss reviewed (commented):
Thanks for opening the PR! I think is looking good so far, but I miss the documentation. Could add it, please? 
The documentation for this would be placed in `website/content/docs/datasources`, you will also have to add them to the navigation in `website/data/docs-nav-data.json`.

sylviamoss commented on datasource/http/data.go:
```diff
@@ -0,0 +1,150 @@
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return cty.NullVal(cty.EmptyObject), err
```
This error will always be null
```suggestion
		return cty.NullVal(cty.EmptyObject), fmt.Errorf("HTTP request error. Response code: %d", resp.StatusCode)
```

sylviamoss commented on datasource/http/data.go:
```diff
@@ -0,0 +1,150 @@
+
+// Most of this code comes from http terraform provider data source
+// https://github.com/hashicorp/terraform-provider-http/blob/main/internal/provider/data_source.go
+func (d *Datasource) Execute() (cty.Value, error) {
```
The returned error message can be improved a little bit to help understand the step it failed.

teddylear replied:
Added some print statements before throwing errors to make them more traceable. But I assume we still want to throw current error up the chain.

sylviamoss commented on website/content/partials/datasource/http/Config-not-required.mdx:
```diff
@@ -0,0 +1,5 @@
+<!-- Code generated from the comments of the Config struct in datasource/http/data.go; DO NOT EDIT MANUALLY -->
```
You can use these generated partials in the docs. See the other datasources documentation as an example.

teddylear replied:
Updated PR with docs

sylviamoss commented on datasource/http/data_acc_test.go:
```diff
@@ -0,0 +1,109 @@
+var testDatasource404Url string
+
+func TestHttpDataSource(t *testing.T) {
+	tests := []struct {
```
Nice!

teddylear replied:
Thanks :)

teddylear commented:
@sylviamoss Updated PR based on feedback. Please let me know if I should update anything else.

sylviamoss commented on website/content/docs/datasources/http/http.md:
```diff
@@ -0,0 +1,49 @@
+
+Type: `http`
+
+The `http` Data Source retrieves information from an http endpoint to be used
+during packer builds
```
Copying and pasting the terraform datasource docs because I think the wording was a little better. What do you think?
```suggestion
The `http` data source makes an HTTP GET request to the given URL and exports information about the response.
```

teddylear replied:
Yeah that sounds better

sylviamoss commented on website/content/partials/datasource/http/Config.mdx:
```diff
@@ -0,0 +1,5 @@
+<!-- Code generated from the comments of the Config struct in datasource/http/data.go; DO NOT EDIT MANUALLY -->
```
Can't find where this one is coming from. Does the CI fail if you remove it? I think it was generated the first time

teddylear replied:
Let me try to remove it and see what happens

teddylear replied:
Looks like website changes passed with removing this.

sylviamoss reviewed (commented):
Thanks for making the changes! I made some suggestions on the docs.

sylviamoss commented on datasource/http/data.go:
```diff
@@ -0,0 +1,153 @@
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+	// Url where should be getting things from
```
Keeping it consistent with the terraform docs
```suggestion
	// The URL to request data from. This URL must respond with a `200 OK` response and a `text/*` or `application/json` Content-Type
```

sylviamoss commented on datasource/http/data.go:
```diff
@@ -0,0 +1,153 @@
+	common.PackerConfig `mapstructure:",squash"`
+	// Url where should be getting things from
+	Url string `mapstructure:"url" required:"true"`
+	// Request headers for call
```
Same as above
```suggestion
	// A map of strings representing additional HTTP headers to include in the request.
```

sylviamoss commented on datasource/http/data.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+type DatasourceOutput struct {
+	Url              string            `mapstructure:"url"`
+	Response_body    string            `mapstructure:"body"`
+	Response_headers map[string]string `mapstructure:"request_headers"`
```
```suggestion
        // The URL the data was requested from. 
	Url              string            `mapstructure:"url"`
	// The raw body of the HTTP response.
	Response_body    string            `mapstructure:"body"`
	// A map of strings representing the response HTTP headers. 
	// Duplicate headers are contatenated with , according to [RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2)
	Response_headers map[string]string `mapstructure:"request_headers"`
```

teddylear commented on website/content/docs/datasources/http/http.md:
````diff
@@ -0,0 +1,49 @@
+  bucket = "hardened-ubuntu-16-04"
+  channel = "production-stable"
+}
+```
````
Ahhh good catch, my bad

teddylear commented:
Updated docs with suggestions (thanks again for suggestions, writing better docs is something I need to improve on). If there's anything else I should change or if you want to to rebase to clean up commit history please let me know.

teddylear commented:
@sylviamoss It looks like the vercel deployment failed but it's unclear why, any ideas?

sylviamoss commented on website/content/docs/datasources/http/http.md:
```diff
@@ -0,0 +1,49 @@
+---
```
The build is failing because the file type should be `mdx`. 

The best place to put this doc is directly under the `datasources` folder, as `datasources/http.mdx` because it's a single datasource type.

sylviamoss commented on website/data/docs-nav-data.json:
```diff
@@ -712,6 +712,10 @@
             "title": "Image",
             "path": "datasources/hcp/hcp-packer-image"
           },
+          {
```
Here, you put the datasource under the HCP datasources. You need to move this one up and update the path to accordingly to the changes I proposed above `datasources/http`

sylviamoss reviewed (commented):
Thanks for the patience! I've been focused on other stuff in parallel and sometimes it takes longer to switch context

teddylear commented:
@sylviamoss Thanks again for all the reviews.I think I addressed issues you mentioned and the container builds and runs locally for me. Please let me know if there's anything else I should change.

teddylear commented on datasource/http/data.go:
```diff
@@ -0,0 +1,157 @@
+	url, headers := d.config.Url, d.config.Request_headers
+	client := &http.Client{}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	// TODO: How to make a test case for this?
+	if err != nil {
+		fmt.Println("Error creating http request")
+		return cty.NullVal(cty.EmptyObject), err
+	}
```
I can make a PR with this update. Thanks!
